app/src/org/commcare/views/UserfacingErrorHandling.java

```diff
@@ -64,6 +72,35 @@ public static void createErrorDialog(final CommCareActivity activity, String err
                     activity.dismissAlertDialog();
                 };
         factory.setPositiveButton(buttonDisplayText, buttonListener);
-        activity.showAlertDialog(factory);
+        return factory;
+    }
+
+    public static StandardAlertDialog getErrorDialog(final FormHierarchyActivity activity, String errorMsg,
```

Comment:
@Charl1996 your code looks right to me though looks like we can optimise the design here to avoid code duplication like this. The way to do it would be to refactor this class to take a generalised parameterised object rather than a concrete class like `CommCareActivity` or `FormHierarchyActivity`. Here is an example of how we can do this - 

1. Create a new interface for alert dialog methods used in this class - 

````
public interface AlertDialogController {

    /**
     * Show the alert dialog provided by the given AlertDialogFactory
     */
    void showAlertDialog(CommCareAlertDialog dialog);

    /**
     * Dismiss the current alert dialog
     */
    void dismissAlertDialog();
}

````

These methods are also present in `DialogController` . So we can modify it by removing these methods from `DialogController` and have it extend `AlertDialogController` 

`public interface DialogController extends AlertDialogController {`



2. Next thing we need to is utilise our new interface `AlertDialogController` in `UserfacingErrorHandling`. It seems best to do it by using a parametrised argument.

````
public class UserfacingErrorHandling<T extends AppCompatActivity & AlertDialogController> {
 
 public void logErrorAndShowDialog(final T dialogActivity,
                                             XPathException exception,
                                             final boolean shouldExit)
}
````
we will need to change methods in this class from `static` to `non static` to use parametrised arguments. That means we will need to modify the existing calls to these functions as - 

`new UserfacingErrorHandling().logErrorAndShowDialog(this, e, true);` 

that is initialise the instance of `UserfacingErrorHandling` before calling it.


Apologies that we are getting into a bit of design refactoring as part of this ticket when it was supposed to be minor work but think it's going to be good learning and intro to some of the Java code design.